### DISCUSSION TITLE
Join us to chat and sync-up about Owncast.

### DISCUSSION THREAD

**Discussion by gabek** [owncast/owncast#167] [General Owncast discussion]
Hey, all!  This is a longish message that came out of a conversation we had Thursday.

To begin with, I had a great time chatting with @graywolf336, @geekgonecrazy and @gingervitis the other day.  We didn't really plan to have a more encompassing "team call", but to everyone else that didn't get notified, we'd love to have you join us if you'd like to next week.  We talked high level about the project, and wanted to detail a list of different potential goals we can focus on in the near-ish term.

## Owncast sync this week

This Thursday, the 24th, [3pm Pacific, 5pm Central](https://www.worldtimebuddy.com/?qm=1&lid=5391959,6,100&h=100&date=2020-9-19&sln=22-23) we'll be chatting again.  So @mattdsteele, @jeyemwey join us if you're interested and you're free.  I created this little link on Eventolist to have an easy way to download an ics. Hopefully that's usable and puts it in your local timezone. https://www.eventolist.com/e/DCFFIO5BLK1L8LDH05ZA.
Meeting Link: https://meet.jit.si/owncast-discovery-meeting

Actionable things I'd like to discuss this week

1. How can we plan for, and leverage Hacktoberfest? (#161)
2. Releasing [0.0.2](https://github.com/owncast/owncast/milestone/2).
3. Spinning up demo instances.
4. Beginning the admin site project. (#114)
5. Discussing the current approach to the directory's YP API. (Low priority background work)

## Goals

As far as goals, I split things into four categories: The **core project**, **awareness + adoption**, **broadcaster community + engagement**, and what's needed to offer a **hosted solution**.  Luckily I believe many of these go hand-in-hand.  This is not encompassing, but hopefully it'll get everyone thinking about what direction and features they'd like to see.

Some of these items I've already been splitting up into the [release milestones](https://github.com/owncast/owncast/milestones) in Github issues.  As some things have dependencies on others, spitting features up in this way seems to make sense.  I encourage you to take a look.

## Core Project

These are the things that I believe are requirements for the project itself to be useful to a wide variety of people, fulfilling the initial vision.

### Web admin (#114)

Have a read-only page for operators to see how their instance is configured and the status of it.  The second step is allowing configuration changes through this admin site, while things like managing recordings, chat moderation, etc will be added as well.

### Recordings (#102)

People expect to have recordings as as core functionality for them.  It has a direct dependency on the web admin as being able to delete, hide, and edit metadata of your recordings will be required.

For the initial version of recordings I don't think we will have chat integration, but for future updates I can see us wanting to replay the chat history synchronized with the video.

I can start documenting some of the specifics around this so we can discuss the pieces in play, because it touches a lot.  So hopefully soon we can break it into actionable chunks of work.

### Scheduled events (#150)

This is useful with both the specific use case that Bradley outlined in his original proposal, but also as engagement between broadcasters and viewers.

Much like recordings, this is a feature that would touch many pieces, so hopefully soon we can break this feature into chunks of work post-initial admin site.

## Awareness and adoption

These are the things that help clarify to people what the goals of the project are, how it can be useful to them, and reduce the number of reasons people may not be interested in trying it.

### Owncast instance directory (#142)

There are two camps of people who live stream: Those who have an audience (conferences, churches, businesses) and those looking to build an audience (hobbyists, people in the arts, performers, online personalities).  For the later crowd a bare minimum answer to the question "how will people discover my content?" is required.  So it really solves two problems: It clarifies what the project is by offering a wide variety of instances for people to visit, and it supports broadcasters looking to increase their reach.  The downside of this directory is it'll be empty until there are more instances.  I have a WIP I've been playing with here and there at https://owncast-web-test.gabek.vercel.app/ and the project (frontend and backend) lives in https://github.com/owncast/owncast-yp.  It'll need polish, but I'll continue getting it functional.

### Demo instance(s)

We don't currently have any instances of Owncast running that people can actively check out.  A 24/7 instance that's streaming some kind of accessible content (like public domain cartoons or or a recording of a previous stream), while not *exactly* like a live stream, since the content isn't live, will give people a feel for it.  I'm also not above having fake chat messages scroll by so people can get the feel of activity.  A secondary benefit of running instances ourselves is it'll start populating the directory, as well.

## Broadcaster community and engagement 

This is a category I don't know how deep, if at all, we want to dive into.  If people look at Owncast as an alternative to Facebook Live, Twitch, YouTube live, etc then there's certain things that people expect. 

### Following & Notifications

Allowing people to "follow" an instance, and get notified of every time a live stream begins would increase the engagement of each individual instance.  This is something that people expect coming from the more social streaming world.  Also, with the future ability to schedule live events (as mentioned above #150) it gives more engagement points when reaching out to viewers (Example: "Only three more hours until wedding/service/party/etc!")

### Monetization

While I believe it's out of scope for this project, it's something to keep in mind that monetization of live streams is really big right now.  And while I don't think we should build a twitch-level payment platform, there may be a middle ground we can find to have built-in linking to existing payment processors, bitcoin, etc.

### Near-term

Instead of a custom infrastructure and deployment solution I propose when we're at the place where we can offer "hosted" for a select few people we just spin up, manually, standalone instances, and run it for them, likely free of charge.  This solution doesn't scale, but for an early experiment we could probably learn a lot from this simple setup.

### Configuration/Data management

- Use a real database server instead of of SQLite for hosted customers.  Our config can probably be expanded upon to support different DSNs.

- Use a database instead of the file system for config, stats, and other misc persisted values. A perfect world scenario would allow for standalone instances to use a config file + sqlite and hosted customers to be completely database driven to allow for scalability and flexible deployment.

- Statically assign a set of default video variants that each customer gets instead of allowing people to worry about and customize it. High, medium and low. 

- We already support external storage of video segments, so we could probably just use a single bucket for bulk storage of video segments (depending on security concerns of such things).

- Chat moderation may be a must-have since hosted customers will be less technical and won’t be taking on the risk themselves.  As a hosting provider we'll be taking on the risk and responsibility of reducing harassment and other possible unwanted chat content.

### Deployment

- Deployment architecture that spins up an instance and knows where to find the config and transcoders for that user. 

- Scalable transcoders that are decoupled from the standalone instance. The web UI and APIs are lightweight and many can be run on a single machine in comparison to the work needed to transcode the video.

**Comment by jeyemwey** [comment]
3pm Pacific is midnight in Central Europe, so I'm not sure if I can make it. But I will try!

**Comment by mattdsteele** [comment]
Yeah I'm in!

**Comment by gabek** [comment]
Looking forward to chatting with you all tomorrow!

**Comment by gabek** [comment]
I'm not sure when is good for everyone else, but to be a bit kinder to @jeyemwey, what earlier time works for all of you next week?  I'm currently pretty flexible.  Can we bump it back maybe a couple hours so he's not up at midnight? :)

**Comment by jeyemwey** [comment]
8pm UTC / 1pm Pacific / 10pm CEST would be great for me, if that is possible.

**Comment by gingervitis** [comment]
Are weekends an option?

**Comment by gabek** [comment]
They work for me.

**Comment by graywolf336** [comment]
Weekends are an option, although generally would have to be on Saturdays or a short time window on Sundays. Usually around 2pm to 4pm, Sundays are one of my busiest days off a computer.

**Comment by gingervitis** [comment]
I won't be able to consistently make our current Thursday afternoon times. 
Sundays are pretty busy for me as well.

I would also be open to doing a Tues or Thurs mornings before 10am Pacific / 12pm CT / 5pm UTC.
I can also make Saturdays work.

**Comment by gabek** [comment]
Not sure if this poll thing is useful (or even if it works at all!) but I'm willing to try it.

Mark your availability!  https://doodle.com/poll/pes8tfa2u6dn8epw

**Comment by gabek** [comment]
Date selected, thanks!

### [This Thursday, 9:30am Pacific, 4:30pm UTC](https://www.worldtimebuddy.com/?qm=1&lid=5391959,6,100&h=100&date=2020-10-1&sln=16.5-17.5)
Meeting Link: https://meet.jit.si/owncast-discovery-meeting

We'll chat about the following:

* Admin backend work is moving forward.  Frontend stuff will start soon.  Probably not much to need to discuss there.
* Broadcasting demo site.  Clarify how we want the stream key rotation to work.  #182 
* Javascript dependencies.  It's come up a couple times now that people aren't comfortable with external sourcing, plus it slows down loading of the site.   How do we move forward and resolve this?  #189 
* Are there small, non-technical/non-feature related things we can start on now as background work to help us get to a hosted Owncast offering?  Product/feature related things are all being worked on, but what about other categories of work that needs to be accomplished in order to offer things to people as a service?
* Discuss releases and testing (manual & automated).  Due to a human error on my part we had a bug that went unseen in the 0.0.2 release.
* Discuss the chat username thing.  Should we prompt for name immediately to make that clearer for people instead of making them change their name later?  We can add url params for pre-populating that name in "invite links".
* Custom paths.  We have some custom public/private storage paths currently available in the config, but we're only marginally supporting them.  I propose we remove these custom paths and come back around to do it differently and better later to allow us to move faster now.

**Comment by mattdsteele** [comment]
Sorry folks, I got pulled into a work meeting this morning and won’t be able to attend.

**Comment by gabek** [comment]
## Let's sync-up next week (Week of Oct 19)

Mark your availability here: https://doodle.com/poll/phmk4pyw87d8xwvt

@mattdsteele @geekgonecrazy @gingervitis @graywolf336 @jeyemwey 
@jmhobbs If you're interested, join us!  Anybody else who sees this and is interested you're invited too!

Some topics:

1. The road to [0.0.3](https://github.com/owncast/owncast/milestone/3).
1. The status of the admin project.
1. Planning for [0.0.4](https://github.com/owncast/owncast/milestone/4) and a read/write, persistent data store.  Discuss the proposed approach, how to migrate to it, how much of it we want to move at once, and what functionality we need to make that happen.
1. The status of the [YP project](https://directory.owncast.online).
1. Continued background planning on what we can accomplish now to move us closer to a hosted solution.  Let's come up with one actionable item we can work towards.
1. Discuss how people are discovering Owncast.  Can we leverage this and learn anything from it?  Can we get feedback from any of these people?

We'll meet up at the regular link: https://meet.jit.si/owncast-discovery-meeting

**Comment by gabek** [comment]
Date and time set!  [Thurs Oct 22 9:15 AM](https://www.worldtimebuddy.com/?qm=1&lid=5391959,6,100&h=5391959&date=2020-10-22&sln=9.25-10.25)

Looking forward to chatting with you all soon!  https://meet.jit.si/owncast-discovery-meeting

See above for some topics we can go over, and bring your own.

**Comment by gabek** [comment]
Recap of some of the things from the call:

* @bitspill and @jmhobbs joined us!
* @bitspill let us know that Owncast doesn't work on Windows.
* @jmhobbs is experimenting with Go Plugins.  But it turns out they don't work on Windows either :)
* The admin project is moving forward.  There's a layout, and charts, and tables.
* Figuring out continued decoupling of the video pipeline from the Owncast instance is the next big step in being able to support a hosted architecture.  We need a transcoder instance to be able to save video without passing it back to the lightweight owncast (web, chat) instance, but owncast will still need to know where to find the HLS stream.  I have some ideas I'll write up that we can discuss.
* @jeyemwey is going to see if he can get the YP project running and help with the frontend.
* I've been working on the persistent config framework and it should be ready to go once we have our first read/write config functionality.

If I missed anything that we chatted about, please post!

**Comment by gabek** [comment]
It's that time again, let's chat about Owncast!

### Let's get 0.0.3 out!

As we're more or less complete with everything we want in [0.0.3](https://github.com/owncast/owncast/milestone/3?closed=1) let's get together to talk about this release.  Testing, documentation, web updates, bugs, tweaks, release plans and anything else on our minds to figure out how and when to get it out.

### Let's plan for 0.0.4!
We can also kick off our plans for [0.0.4](https://github.com/owncast/owncast/milestone/4) and talk about the next version of server configuration using the web admin.

### Pick a time: https://doodle.com/poll/exxurbrr9ayn5gt2

I know there's the holiday for many of us next week, so scheduling might be weird.  If these options don't work at all for people let me know.

**Comment by gabek** [comment]
We're good to go for December 1st!  [9:30am Pacific, 11:30am Central, 5:30p UTC](https://www.worldtimebuddy.com/?qm=1&lid=5391959,6,100&h=5391959&date=2020-12-1&sln=9.5-10.5&hf=1).

**Reply by tian2992** [reply]
Hello. I just found out about this and I am looking forward to participate :) nice meeting you all :)

**Reply by tian2992** [reply]
P.S. I think i broke the demo stream, as after i streamed a bit, it started returning Code 502 :S

**Reply by gabek** [reply]
> P.S. I think i broke the demo stream, as after i streamed a bit, it started returning Code 502 :S

The `broadcast.owncast.online` server?  Interesting!  What were you using to stream?  Do you have any specific configuration you can share with us?

**Reply by tian2992** [reply]
i was just trying out straming with OBS Studio 21.1 on windows x64. idk if it is something of use on the logs. I am looking to mount my server on a DO w/2 core for testing tomorrow.

**Comment by gabek** [comment]
I look forward to chatting with you all on Tuesday, [9:30am Pacific, 11:30am Central, 5:30p UTC](https://www.worldtimebuddy.com/?qm=1&lid=5391959,6,100&h=5391959&date=2020-12-1&sln=9.5-10.5&hf=1) at https://meet.jit.si/owncast-discovery-meeting.

Here's some things I'd like to put on the docket for us to discuss.

* Game plan for [releasing 0.0.3](https://github.com/owncast/owncast/issues/399).  We have a little demo video now and we have updated documentation ready to go.  Who should we tell about it?  @muesli had some thoughts about how we should bubble this up a bit.
* Kicking off [0.0.4 development](https://github.com/owncast/owncast/milestone/4).  I have a branch with the start of the data persistence layer and I'm wiring it up to a couple simple APIs.  As a reminder, this will be the work to support configuration of an owncast instance from the web instead of the config file.
* Status and thoughts of plugins for storage, integrations, analytics, etc.  We last discussed using something like [go-plugin](https://github.com/hashicorp/go-plugin) due to the downsides @muesli outlined in native Go plugins.
* [Unique viewer counts](https://github.com/owncast/owncast/discussions/374) are currently limited to only people watching video via the Owncast web interface.  Anybody have any ideas to do this better?
* We have a design for the in-development [Owncast Directory!](https://directory.owncast.online)  I think we're getting closer to publicly launching this project.  For now, however, if you can [enable YP support in your config file](https://directory.owncast.online/get-listed) (0.0.3/master) some testing and making sure instances show up as expected would be a help.  For 0.0.3 we can support the option to enable it (off by default) and if people feel like turning it on then that's cool.

Bring whatever else is on your mind.  Talk to you Tuesday at https://meet.jit.si/owncast-discovery-meeting.

**Reply by mattdsteele** [reply]
Was hoping to make this, but work got in the way. Let me know if there's anything I can help with!

**Reply by geekgonecrazy** [reply]
Same here 🙁 hopefully was productive call

**Comment by brown7477** [comment]
Hey There, Gabe told me to check in here to get some information on OwnCast.  I am kind of excited to learn about it.

**Reply by gabek** [reply]
Hey there!  This thread is actually about scheduling meeting times.  Would you mind creating a new topic with any questions you have?

**Comment by gabek** [comment]
Exciting times!  Let's get together next week and chat about the fantastic response we've had since the 0.0.3 launch and continue planning out our next steps!  Mark your availability here: https://doodle.com/poll/gzspintzerrhmisg

As you can see by taking a look at our [current milestone roadmap](https://github.com/owncast/owncast/milestones) some of the priorities and plans have been moved around a bit.  Take a look at that schedule and come prepared to talk about some of these topics:

* Recent buzz and interest.  What went well?  Is there anything we can learn going forward when talking about the project?
* [Chat moderation](https://github.com/owncast/owncast/milestone/8).  What's the scope?
* [ActivityPub/Fediverse integration](https://github.com/owncast/owncast/issues/458).
* It's time to solidify the plans updated [admin to support config updates](https://github.com/owncast/owncast/issues/304), so if you have any specifics or ideas around that we should start to go over them.
* Future of chat.  Most requests and enhancements going forward are going to be around chat.  It's likely 90% of work going forward with features, complexity, UI and requests will be around just chat and user interaction related items.  How do we handle that?
* A PR for autoplay (#495) was put up, but has some caveats.  Let's discuss.
* Or anything else!

**Comment by gabek** [comment]
We're locked in to chat on Monday!  [1:30p Pacific, 3:30 Central, 10:30 CET](https://www.worldtimebuddy.com/?qm=1&lid=5368361,5128581,12,6&h=5368361&date=2020-12-28&sln=13.5-14.5&hf=1).

Come prepared to talk [about the above](https://github.com/owncast/owncast/discussions/167#discussioncomment-228568) items, along with anything else on your mind!

We'll be using our usual Jitsi meeting link: https://meet.jit.si/owncast-discovery-meeting

I look forward to syncing back up, reflecting on some of the fantastic feedback we've received lately, and kicking off some new features.

See you Monday!

**Reply by aaronpk** [reply]
Is this just for maintainers or is it open to others? I am potentially interested in joining too :)

**Reply by gabek** [reply]
Totally open to everybody, we'd love to have you!

**Comment by gabek** [comment]
## Dec 28 Summary

- 0.0.4 is about ready to go.  Just some sanity testing left.  Within the next day or so I'll push it out.  Any additional testing on `master` would be appreciated.
- We discussed ActivityPub and as a group reiterated it's worth the development time since it'll open up functionality and features in the future.
- Discussed the status of the update to the admin with the full configuration UI, moving away from the config file.  A possible importer/exporter between config files and the new config database was seen as something that could be useful, even just as a CLI tool.
- Getting documentation in front of people has been challenging.  We decided that adding more links to documentation into the admin itself would be a good next step since people will be spending more time in the admin going forward.  A "help" page with links to troubleshooting documents may be helpful.
- We decided to not act upon making any changes to facilitate AutoPlay right now, since there are caveats around the experience.  In the future a possible admin toggle to enable/disable AutoPlay might be nice, but there's UI considerations on the player if we ever want to go down that road.
- Similarly, we won't be adding a feature to password protect the stream at this moment.  If we get more demand in the future we can re-address.  

### Discussion about chat

User message-based chat moderation is in a branch.  You can hide messages, and re-show them.  It's basic but will give people control over their chat content for the first time.

Big picture we discussed how deep into being a fully fledged chat application we want to go. People want accounts, 3rd party auth, and tool integrations.  It was suggested we focus on building interfaces for other people to build the functionality they want on top of Owncast.  Bots, event handlers, integrations, etc.  We can never be everything to everyone, and we're not a full-time chat development team, but we can focus on hooks and APIs so people can do some really fun and useful things with our chat.

But we do need to tighten down our identity first, so we discussed at a high level [this idea](https://github.com/owncast/owncast/discussions/440) for assigning identity without requiring authentication as a way to handle user moderation (as opposed to message hiding-based moderation).

We also discussed what we will need to do to lock down our web socket, and it was suggested maybe a CSRF token, and limit the web socket to the web app only, and allow future integrations to use specific 3rd party APIs, so we can keep the websocket locked down.

**Comment by aaronpk** [comment]
Sorry I haven't been able to join yet, are you still doing these calls? Would love to join if the timing works out next!

**Reply by gabek** [reply]
We had one last week, but it was organized and scheduled on Rocket.Chat.  I'll make sure to continue to post them here going forward.  We're going to try and do them monthly from now on and be a bit more consistent, and we can start to plan them a bit further in advance.  I'll post the notes from the call here.

**Reply by aaronpk** [reply]
Thanks! I didn't realize there was a rocket.chat too

**Comment by gabek** [comment]
Here's the notes from our March 18th call.

@nebunez put together some slides on some topics he wanted to go over, and those can be found [here](https://nextcloud.w0rmh0le.net/index.php/s/EJQzeSeccZXSqEa?dir=undefined&openfile=417683).

- We still obviously really want ActivityPub integration, but compared to doing recordings first recordings would be more impactful, increase engagement, and fulfil a real need for streamers.
- However, there's an application for an NLnet grant in to fund ActivityPub development, so if that were to get accepted (unlikely!) we could bump that up.
- Discussed embedding a single currently live stream in the directory to show off that live things are happening and renaming the directory to owncast.tv.
- Nebunez thought a lot about extensions, core vs. external features, and how non-core functionality that can fundimentally alter the UI and core experience can exist for people. The idea of a slightly-official fork of Owncast that has more experimentation and non-core features was floated.
- He also brought up the idea of an "extention hub", an external interface where you can register additional functionality that talk to the hub, and the hub ties into Owncast.
- See his slides for more details. A middle ground was discussed where people should be encouraged to prototype more and try out ideas on their servers, and we should continue to think of generic ways to add integration points to Owncast for things that people want to do, including for page UI. The external actions modal is an example of that.
- Is there a way we can reach people who want to stream using Owncast that we can offer up some free instances to. This would increase the number of streamers leading to more feedback, testing, outreach, etc. Could fund with donations that come in. Would also be a testing ground for any possible hosted service in the future.
- Discussed getting Owncast more involved in the streaming world. Is there a way for us to host discussions, Q/As, other streamers (from Twitch, Youtube), people developing other platforms (Peertube?) to talk about the streaming ecosystem and force ourselves into the conversation?
- Keep an eye out for somebody who is into streaming who might want to play a role with owncast as a "community manager". This could go hand in hand with somebody we run a free instance for.
- Discussed git, releases and versioning. We decided to version as 0.1.0 once we have recordings and can make a big splash about the version bump and new features.
- Also discussed a possible move to gitflow, but thought maybe just renaming the default branch from master to "develop" might be good enough for now, to let people know master is not the same as a stable release.
- We should have calls more often, we'll try monthly going forward.

**Comment by gabek** [comment]
Scheduling our next call for later this month.  Please select a time that works for you.  If nothing works let me know, and we can figure out some other options so those who want to attend can.  https://doodle.com/poll/g4e76mm4nbhu7276?utm_source=poll&utm_medium=link

**Reply by gabek** [reply]
Notes from our last call got posted here: https://github.com/owncast/owncast/discussions/966

**Comment by gabek** [comment]
Let's get together before June is over and chat!  Mark your availability here:

https://doodle.com/poll/btzzpprvxn7v8qfb

**Reply by gabek** [reply]
Sorry, scheduling URL changed: https://doodle.com/poll/btzzpprvxn7v8qfb

**Reply by gabek** [reply]
Notes from this call were saved to: https://collab.owncast.tv/code/#/2/code/view/Mhf7jsNaPvMQLLXcS4jn7qCK0b05HoxTjKWDN7Flg3w/

**Comment by gabek** [comment]
The next public Owncast sync-up chat will be [June 24th](https://www.timeanddate.com/worldclock/converter.html?iso=20210624T200000&p1=224&p2=1440&p3=tz_et&p4=tz_cest) and you're invited!  Join us on [Jitsi](https://meet.jit.si/owncast-discovery-meeting) at [1:00PM Pacific, 10PM CEST, 20:00 UTC](https://www.timeanddate.com/worldclock/converter.html?iso=20210624T200000&p1=224&p2=1440&p3=tz_et&p4=tz_cest).

**Reply by gabek** [reply]
Notes from this call were saved to: https://collab.owncast.tv/code/#/2/code/view/Mhf7jsNaPvMQLLXcS4jn7qCK0b05HoxTjKWDN7Flg3w/

**Comment by gabek** [comment]
Our next call will be coming up.  Please mark your availability so we can hang out and talk about Owncast!

https://doodle.com/poll/cemfaaeywrtddmh3

**Reply by gabek** [reply]
Our next Owncast call will be [Thursday, the 19th](http://www.worldtimebuddy.com/event?lid=12%2C8%2C5391959&h=12&sts=27155400&sln=22-23&a=show&euid=326725ad-e5a9-8294-c9a5-153724504b97).  I'll be putting some topics together for us to discuss, so [please take a look](https://collab.owncast.tv/code/#/2/code/view/AphEJXNB0Fjwgj-djde7AwjwUsbpaBwkHT+YMg8TWt4/) beforehand, though these topics will be incomplete and in progress up until we get on the call.
